swift/apple/FirezoneNetworkExtension/Adapter.swift

```diff
@@ -313,71 +226,109 @@ extension Adapter {
+  /// cases most commonly:
+  /// - New DNS servers were set by DHCP
+  /// - The user manually changed the DNS servers in the system settings
+  ///
+  /// For now, this will break DNS if the old servers connlib is using are no longer valid, and
+  /// can only be fixed with a sign out and sign back in which restarts the NetworkExtension.
+  ///
+  /// On macOS, Apple has exposed the SystemConfiguration framework which makes this easy and
+  /// doesn't suffer from this issue.
+  ///
+  /// See the following issues for discussion around the above issue:
+  /// - https://github.com/firezone/firezone/issues/3302
+  /// - https://github.com/firezone/firezone/issues/3343
+  /// - https://github.com/firezone/firezone/issues/3235
+  /// - https://github.com/firezone/firezone/issues/3175
   private func didReceivePathUpdate(path: Network.NWPath) {
-    // Will be invoked in the workQueue by the path monitor
-    switch self.state {
+    if case .tunnelReady(let session) = state {
+      // Only respond to path updates if the tunnel is up and functioning
 
-    case .startingTunnel(let session, let onStarted):
-      if path.status != .satisfied {
-        self.logger.log("Adapter.didReceivePathUpdate: Offline. Shutting down connlib.")
-        onStarted?(nil)
-        resetToSystemDNS()
-        self.packetTunnelProvider?.reasserting = true
-        self.state = .stoppingTunnelTemporarily(session: session, onStopped: nil)
-        session.disconnect()
-      }
+      if path.status == .unsatisfied {
+        logger.log("\(#function): Detected network change: Offline.")
 
-    case .tunnelReady(let session):
-      if path.status != .satisfied {
-        self.logger.log("Adapter.didReceivePathUpdate: Offline. Shutting down connlib.")
-        resetToSystemDNS()
-        self.packetTunnelProvider?.reasserting = true
-        self.state = .stoppingTunnelTemporarily(session: session, onStopped: nil)
-        session.disconnect()
-      }
+        // TODO: Tell connlib we're offline so it can sleep its retries?
 
-    case .stoppingTunnelTemporarily:
-      break
+        // Check if we need to set reasserting, avoids OS log spam and potentially other side effects
+        if packetTunnelProvider?.reasserting == false {
+          // Tell the UI we're not connected
+          packetTunnelProvider?.reasserting = true
+        }
+      } else {
+        self.logger.log("\(#function): Detected network change: Online.")
+
+        // Hint to connlib we're back online, but only do so if our primary interface changes,
+        // and therefore we need to bump sockets. On darwin, this is needed to send packets
+        // out of a different interface even when 0.0.0.0 is used as the source.
+        // If our primary interface changes, we can be certain the old socket shouldn't be
+        // used anymore.
+        if (path.availableInterfaces.first?.name != primaryInterfaceName) {
```

Comment:
@thomaseizinger @conectado I only call reconnect if the primary interface changes.

Reply:
Great!